enn_zoo/enn_zoo/vizdoom_env/vizdoom_environment.py

```diff
@@ -0,0 +1,242 @@
+
+        # We do not need screen stuff so minimize things (apart from rendering)
+        self._doomgame.set_window_visible(False)
+        self._doomgame.set_screen_format(vzd.ScreenFormat.CRCGCB)
+        self._doomgame.set_screen_resolution(vzd.ScreenResolution.RES_160X120)
+        self._image_width = 160
+        self._image_height = 120
+
+        # But we need list of objects
+        self._doomgame.set_objects_info_enabled(True)
+
+        # Handle GameVariables. Add the ones we force
+        self._game_variables = self._doomgame.get_available_game_variables()
+        self._game_variables = list(
+            set(self._game_variables).union(set(FORCED_GAME_VARIABLES))
+        )
+        self._doomgame.set_available_game_variables(self._game_variables)
+        self._game_variable_names = [
+            game_variable.name for game_variable in self._game_variables
+        ]
+        # Rename variables where needed
+        for original_name, new_name in RENAME_GAME_VARIABLES.items():
+            if original_name in self._game_variable_names:
+                self._game_variable_names[
+                    self._game_variable_names.index(original_name)
+                ] = new_name
+        # Find where angle and pitch are so we can turn them into radians
+        self._angle_index = self._game_variable_names.index("angle")
+        self._pitch_index = self._game_variable_names.index("pitch")
+
+        # When episode terminates the buffers may be empty, so instead we return the last valid observation
+        self._last_observation: Observation = None  # type: ignore
+        self._last_state = None
+        self._episode_steps = 0
+        self._sum_reward = 0
+
+        self._observation_space = ObsSpace(
+            {
+                "Player": Entity(self._game_variable_names),
+                "Objects": Entity(
+                    # TODO "type" here will be just the ord(name[0])
+                    ["x", "y", "z", "type"]
+                ),
+            }
+        )
+
+        self._action_space: Dict[
+            str, Union[CategoricalActionSpace, SelectEntityActionSpace]
+        ] = {}
+        # You can always execute all actions in all steps, so we create a static mask thing
+        self._action_mask = {}
+        self._available_buttons = self._doomgame.get_available_buttons()
+        assert (
+            len(self._available_buttons) > 0
+        ), "No available buttons to the agent. Double-check your config files."
+        for action in self._available_buttons:
+            action = action.name
+            if "DELTA" in action:
+                self._action_space[action] = CategoricalActionSpace(
+                    DELTA_SPEED_STEPS_STRS
+                )
+            else:
+                self._action_space[action] = CategoricalActionSpace(["off", "on"])
+            self._action_mask[action] = CategoricalActionMask(actor_ids=[0], mask=None)
+
+    def obs_space(self) -> ObsSpace:  # type: ignore
```

Comment:
The reason it doesn't like this is probably that the parent method is a `classmethod`. You can just remove this and the action_space method implementations and replace all references with `self.__class__.action_space/obs_space`.

Reply:
![image](https://user-images.githubusercontent.com/5555347/159067829-eebd5d17-3de3-4f2a-a497-dca86ff5921e.png)

Do you mean like this?